API for KEP-4831: Control VPA eviction behavior based on scaling direction and resource

voelzmo opened the pull request:
#### Which component this PR applies to?

<!--
Which autoscaling component hosted in this repository (cluster-autoscaler, vertical-pod-autoscaler, addon-resizer, helm charts) this PR applies to?
-->
vertical-pod-autoscaler

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature
/kind api-change

#### What this PR does / why we need it
Implements KEP-4831: Control VPA eviction behavior based on scaling direction and resource

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
API for #4831
Implementation is done in #5599


#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

voelzmo commented:
While looking for the best place to put this into the Updater, I found https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/updater/priority/pod_eviction_admission.go which was pulled in years ago with https://github.com/kubernetes/autoscaler/pull/1002 but apparently never used afterwards. Beata (after merging) even commented that the default implementation isn't even used: https://github.com/kubernetes/autoscaler/pull/1002#pullrequestreview-130869322

I guess this is probably a good place to implement the filtering we want to do based on the comparison of `target` and `requests` and if no restrictions are given, just use the already implemented `noopPodEvictionAdmission`?

jbartosik commented:
> While looking for the best place to put this into the Updater, I found https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/updater/priority/pod_eviction_admission.go which was pulled in years ago with #1002 but apparently never used afterwards. Beata (after merging) even commented that the default implementation isn't even used: [#1002 (review)](https://github.com/kubernetes/autoscaler/pull/1002#pullrequestreview-130869322)
> 
> I guess this is probably a good place to implement the filtering we want to do based on the comparison of `target` and `requests` and if no restrictions are given, just use the already implemented `noopPodEvictionAdmission`?

Yes, adding a new type that implements `PodEvictionAdmission` and passing it to `Updater` seems to be the best way to implement this KEP.

I'd implement it so when there are no restrictions it just allows the pod

> 
> There's also https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go but that seems to be concerned with limiting eviction based on minimum available replicas and ensuring this also works for Jobs and other Pod owners in the same way.
> 
> Joachim WDYT, should I just make use of the PodEvictionAdmission linked above? From a flow perspective, this seems to be a good fit, [as it is used in the priority calculator](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator.go#L156), where Pod requests and target recommendation are compared to figure out if a Pod should be evicted and in which order.

Yes,

voelzmo commented:
@jbartosik Can you please take a look? Thanks!

voelzmo commented:
@jbartosik Can you try to give this a review, please? Thanks!

jbartosik reviewed (commented):
Since in-place pod updates are now finally a thing (alpha but I hope for beta soon) I'd like to make sure this works well with support for in place updates.

I can see two approaches:

- have one "actuationRequirements" that affect both in-place and eviction actuation
- have separate controls for "evictionRequirements" and "inPlaceUpdateRequirements"

I lean towards single "actuationRequirements" because it's simpler and I don't see how I'd use separate controls for eviction and in place update. What do you think?

jbartosik commented on vertical-pod-autoscaler/deploy/vpa-v1-crd-gen.yaml:
```diff
@@ -3,7 +3,7 @@ kind: CustomResourceDefinition
 metadata:
   annotations:
     api-approved.kubernetes.io: https://github.com/kubernetes/kubernetes/pull/63797
-    controller-gen.kubebuilder.io/version: v0.4.0
```
> 

I sent https://github.com/kubernetes/autoscaler/pull/5754

jbartosik commented:
Let's decide if we want to change the API for in-place updates before merging this (https://github.com/kubernetes/autoscaler/pull/5754)

voelzmo commented:
Force-pushed an update:
* rebased on current master
* removed the `OrEqual` ChangeRequirements as discussed in https://github.com/kubernetes/autoscaler/pull/5754#issuecomment-1647577540

jbartosik commented on vertical-pod-autoscaler/deploy/vpa-v1-crd-gen.yaml:
```diff
@@ -379,6 +379,38 @@ spec:
+                        resource:
+                          description: Resources is a list of one or more resources
+                            that the condition applies to. If more than one resource
+                            is given, they are combined with OR, not with AND.
```
```suggestion
                            is given, requirement is fulfilled if at least one resource meets `changeRequirement`
```

voelzmo replied:
done, thanks!

jbartosik reviewed (commented):
Looks good to me

@mwielgus @gjtempleton could you take a look please

mwielgus reviewed (approved):
/lgtm
/approve
